Default discussion topic

replaceafill opened the pull request:
This PR adds the ability to configure the default topic for new posts in the Discussion tab by setting a new attribute in the Discussion Topic Mapping field.

[Corresponding documentation PR](https://github.com/edx/edx-documentation/pull/1488)

**Dependencies**: None

**Screenshots**:
![studio](https://user-images.githubusercontent.com/560781/26912117-a965c4dc-4bcf-11e7-823c-4d62cf2892a7.png)
![lms](https://user-images.githubusercontent.com/560781/26912118-ae2d9684-4bcf-11e7-96cb-407b320ef306.png)


**Sandbox URL**:

LMS: https://oc2476.sandbox.opencraft.hosting/
Studio: https://studio-oc2476.sandbox.opencraft.hosting/

**Partner information**: 3rd party-hosted open edX instance

**Deployment targets**: edx.org and edge.edx.org

**Merge deadline**: None

**Testing instructions**:

1. In Studio go to the demo course advanced settings.
2. Add a new category to the Discussion Topic Mapping with a "default" attribute set to true. For example:

```json
"Custom": {
    "id": "custom",
    "default": true
}
```

3. Save the changes
4. In the LMS go to the demo course Discussion tab.
5. Add a new post and check that the Topic area default value matches the setting.

**Author notes and concerns**:

We tried to add a JS spec test for the DiscussionBoardFactory change, but the DiscussionRouter interferes with the history of other spec tests. That's the reason why [the current test for the factory is disabled](https://github.com/replaceafill/edx-platform/blob/bfbc3241b396ec9591674983c176839cb6f46aeb/lms/djangoapps/discussion/static/discussion/js/spec/discussion_board_factory_spec.js#L66). 

**Reviewers**
- [ ] @pomegranited 
- [ ] @bdero

edx-webhook commented:
Thanks for the pull request, @replaceafill! I've created [OSPR-1739](https://openedx.atlassian.net/browse/OSPR-1739) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](http://open.edx.org/sites/default/files/wysiwyg/individual-contributor-agreement.pdf)     or indicated your institutional affiliation.     If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

pomegranited commented:
jenkins run quality

bdero commented:
:+1:

@replaceafill Can you open a PR against the edx-documentation repo adding a note about this setting? I think [this section](https://github.com/edx/edx-documentation/blob/e9dd88b6eee7fcd6f5de3e75c7ac18a9cd7e26fc/en_us/shared/manage_live_course/discussions.rst#creating-discussion-topics-for-your-course) would be the most appropriate.

Also, go ahead and add yourself to the [AUTHORS file](https://github.com/edx/edx-platform/blob/master/AUTHORS) if you'd like!

- [x] I tested this: I checked out this branch, added a "Custom" topic, set the defauly discussions topic to "custom", and observed that the default selected topic for posts was "custom".
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~ N/A, this additional setting feeds a default and should have no effect on accessibility.
- [ ] Includes documentation (added note above about this)

@gsong Would you happen to know how we can get Jenkins builds enabled for this PR? Thanks!

gsong commented:
Jenkins ok to test

gsong commented:
@pomegranited @bdero I imagine @replaceafill is covered by OpenCraft's CLA?

bradenmacdonald commented:
@gsong Yes. He's currently a candidate doing some work for us during a trial period, so it's under our CLA.

replaceafill commented:
Thanks @bdero. I created the [PR](https://github.com/edx/edx-documentation/pull/1488) against the edx-documentation repo.

cahrens commented:
Hi @bradenmacdonald and @replaceafill. Sorry for the slow review time, but product and I iterated on a few different possibilities for this feature. We have landed on accepting the feature, but would like to see the following change-- instead of introducing a new setting for this, can you please extend the existing "Discussion Topic Mapping" JSON so that you can specify a topic as the default in that location? That would avoid introducing a new field/setting, keeping all the configuration in one place. So it would look something like (note that if "default" is not present, it is assumed to be false):

````
{
    "Custom Default Topic": {
        "id": "course",
        "default": "true"
    },
    "General": {
        "id": "course"
    },
}

bradenmacdonald commented:
@cahrens Thanks for the acceptance! I think that makes sense - we'll discuss and plan to finish this up next week.

edx-webhook commented:
Hello @replaceafill: We are unable to continue with review of your submission at this time. Please see the associated JIRA ticket for more explanation.

replaceafill commented:
@bradenmacdonald @cahrens Ouch. I just tried closing/reopening this PR to get jenkins to run, but the bot reacted differently than when I did it here: https://github.com/edx/edx-platform/pull/15095

Is there any way to fix it? Sorry about that.

bradenmacdonald commented:
@replaceafill Looks like it worked as far as Jenkins goes - we'll just need @gsong to please re-open the OSPR JIRA ticket. OSPRs follow a different process than the work we do for edX.

replaceafill commented:
@bradenmacdonald Ah. Thanks for explaining that. I hadn't seen the red X next to the commit. I don't understand the bokchoy error though.

bradenmacdonald commented:
Re-opening

bradenmacdonald commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -293,15 +293,10 @@ class CourseFields(object):
-            "the Discussion page."
+            'supported are underscore, hyphen, and period. You can also specify a CategoryName as the default '
+            'for new posts in the Discussion page by setting its "default" attribute to "true". For example, '
+            '"Lydian Mode": {"id": "i4x-UniversityX-MUS101-course-2015_T1", "default": "true"}.'
```
Shouldn't we be using JSON boolean true and not a string "true" ?

```json
{"id": "i4x-UniversityX-MUS101-course-2015_T1", "default": true}
```
vs
```json
{"id": "i4x-UniversityX-MUS101-course-2015_T1", "default": "true"}
```

bradenmacdonald replied:
CC @cahrens in case you have a preference here since I see this was quoted in your example.

replaceafill replied:
@bradenmacdonald Ah yes. A boolean makes more sense there. The Advanced Settings view validates it.

replaceafill commented:
@bradenmacdonald Rebased from d3a785f to resolve conflict

bradenmacdonald commented:
@replaceafill Thanks. Will you be changing the `"true"` to just a boolean `true` as discussed? I'm waiting for that before I can give +1.

replaceafill commented:
@bradenmacdonald Sorry for the delay. I just changed the implementation to use a boolean.

bdero commented:
jenkins run bokchoy

replaceafill commented:
@bradenmacdonald @bdero Rebased this from b94a7eb to resolve a conflict. I didn't understand if [these bokchoy errors](https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/38263/testReport/) were related to my changes or how to fix them.

bdero commented:
> I didn't understand if these bokchoy errors were related to my changes or how to fix them.

Yeah I believe they were failing for everyone and [disabled](https://github.com/edx/edx-platform/pull/15309) shortly later.

bdero commented:
:+1: these changes look good to me. :) @cahrens can you give this another look and remove "Waiting on Author"?

cahrens commented:
The code looks good, but please add unit test coverage.

replaceafill commented:
@cahrens I extracted the code for getting the topic id and added a test case for it. I wrote the reason why I didn't add a spec test under **Author notes and concerns** in the PR description.

cahrens commented:
Thank you, @replaceafill. I will remove the "Waiting for Author" label, and we can schedule the review into our next sprint.

pomegranited commented:
jenkins run js

cahrens commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -293,7 +293,10 @@ class CourseFields(object):
             'category may be "Lydian Mode": {"id": "i4x-UniversityX-MUS101-course-2015_T1"}. The "id" '
             'value for each category must be unique. In "id" values, the only special characters that are '
-            'supported are underscore, hyphen, and period.'
+            'supported are underscore, hyphen, and period. You can also specify a CategoryName as the default '
```
@srpearce Are you OK with the usage of "CategoryName" here? I think it might read better as "You can also specify a category as the default...".

srpearce replied:
I'd prefer to use @cahrens' suggestion.

replaceafill replied:
@cahrens @srpearce Changed.

cahrens commented on lms/djangoapps/discussion/views.py:
```diff
@@ -447,6 +453,7 @@ def _create_discussion_board_context(request, course_key, discussion_id=None, th
         'upgrade_link': check_and_get_upgrade_link(request, user, course.id),
         'upgrade_price': get_cosmetic_verified_display_price(course),
         # ENDTODO
+        'discussion_default_topic_id': _get_discussion_default_topic_id(course),
```
Could you add a comment that None is an acceptable value here, in which case the front-end code will look for a topic that starts with "General" to use as the default? It took me some time to look through the code to understand what would happen if _get_discussion_default_topic_id did not return an entry ID.

replaceafill replied:
@cahrens I added a comment for this. I used 'that contains "General"' instead of 'starts with'.

cahrens commented:
I had a question about the doc string, and a request for a code comment. Otherwise, 👍

replaceafill commented:
@cahrens I addressed your observations. Let me know if I should squash my commits before merging.

cahrens commented:
@replaceafill The changes look good-- go ahead and squash (and then merge after a clean build). Thanks!

replaceafill commented:
@cahrens I squashed this from 5646b2c and the build is clean. Could you please merge it? I don't have permission to do so. Thanks for all your help!

gsong commented:
@replaceafill Thank you for the PR, 🎆.
